Stop emitting <requireLicenseAcceptance> with the default value

jnm2 opened the pull request:
## Bug

Fixes: NuGet/Home#5133
Regression: No

## Fix

Details: `<requireLicenseAcceptance>false</requireLicenseAcceptance>` is now never emitted because `false` is specified to be the default. This seems like the most intuitive starting point.

## Testing/Validation

Tests Added: Yes
Validation: Integration tests failed before fix, passed after fix

jnm2 committed:
Failing tests that <requireLicenseAcceptance> is not emitted when unspecified or default

jnm2 commented:
Also: if you'd like me to take a different approach or fit other constraints, I'd be glad to try.

nkolev92 commented:
There's some failing tests: 

NuGet.CommandLine.Test.NuGetSpecCommandTests.SpecCommand_NoProjectFile
NuGet.CommandLine.Test.NuGetSpecCommandTests.SpecCommand_NoProjectFile_WithArg
NuGet.CommandLine.Test.NuGetSpecCommandTests.SpecCommand_WithProjectFile

My only concern is that this change does not round trip a nuspec. 

Specifically if someone specifies 
```xml
<requireLicenseAcceptance>false</<requireLicenseAcceptance>
```

we should try to honor it. 

Unsure how big of a concern this would be for customers though. 

Thoughts @NuGet/nuget-client @joelverhagen

joelverhagen commented:
This seems like a great change, but there are some concerns I have that should be investigated/mitigated/dismissed:

1. The default is not documented as far as I know: [requireLicenseAcceptance](https://docs.microsoft.com/en-us/nuget/reference/nuspec#requirelicenseacceptance)
    - At the very least, we should document the default so docs match implementation.
1. Was the default _always_ false? For example, in NuGet 2.x/3.x/4.x was it like this?
1. What do other tools do? Do they use the same default?
   - nuget.org
   - NuGet Package Explorer
   - others?

Maybe we can consider an [announcement](https://github.com/NuGet/Announcements) about this clarification as well so if someone comes out of the woodwork after we make this change saying their code has a different interpretation of a default we can point them to a published explanation.

dominoFire commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs:
```diff
@@ -55,7 +55,6 @@ public void PackCommand_IncludeExcludePackageFromNuspec()
     <title>packageA</title>
     <authors>test</authors>
     <owners>test</owners>
-    <requireLicenseAcceptance>false</requireLicenseAcceptance>
```
- Revert changes on those tests. Those represent behaviors NuGet must support. 
- Add tests for testing the generated license acceptance.

dominoFire commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/ManifestTest.cs:
```diff
@@ -309,7 +309,6 @@ public void ReadFromDoesNotThrowIfValidateSchemaIsFalse()
     <version>1.0</version>
     <authors>Luan</authors>
     <owners>Luan</owners>
-    <requireLicenseAcceptance>false</requireLicenseAcceptance>
```
Revert changes.

dominoFire commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageBuilderTest.cs:
```diff
@@ -123,7 +123,6 @@ public void CreatePackageWithNuspecIncludeExcludeAnyGroup()
     <version>1.0.0</version>
     <authors>testAuthor</authors>
     <owners>testAuthor</owners>
-    <requireLicenseAcceptance>false</requireLicenseAcceptance>
```
Revert changes.

dominoFire commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/ProjectFactoryTest.cs:
```diff
@@ -605,7 +605,6 @@ private static string GetNuspecContent()
         <title />
         <authors>Author</authors>
         <owners />
-        <requireLicenseAcceptance>false</requireLicenseAcceptance>
```
Revert changes.

dominoFire commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -4260,5 +4260,129 @@ private void ValidatePackIcon(ProjectFileBuilder projectBuilder)
+                    var xml = XDocument.Load(stream);
+                    ProjectFileUtils.SetTargetFrameworkForProject(xml, "TargetFramework", "netstandard1.4");
+
+                    ProjectFileUtils.AddProperty(xml, "PackageRequireLicenseAcceptance", "true");
```
Add test when PackageRequireLicenseAcceptance = false

jnm2 replied:
That's the test that precedes this one.

dominoFire commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs:
```diff
@@ -4260,5 +4260,129 @@ private void ValidatePackIcon(ProjectFileBuilder projectBuilder)
+        [PlatformTheory(Platform.Windows)]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void PackCommand_RequireLicenseAcceptanceNotEmittedWhenUnspecified(bool withLicense)
```
Merge the tests in one method. Creating a new project from those tests is somewhat time consuming. It is better to modify the project file when needed and make the required assertions.

dominoFire reviewed (dismissed):
Thank you for your contribution.

There's a concern we need to validate before merging this PR:

https://github.com/NuGet/Home/issues/5133#issuecomment-598525205

Another technical concern is that you need to modify `ÌPackageSearchMetadata` to add `bool? requireLicenseAcceptance`. That represents a big breaking change. A workaround for this is to remove the attribute afther the nuspec is created.

Additionally, these are the tests cases we want to validate:

Specified   License? | RequireLicenseAcceptance | Generated   RequireLicenseAcceptance (actual) | Generated   RequireLicenseAcceptance (PR) | Confirmation to install ? | Clients show license?
-- | -- | -- | -- | -- | --
Yes | TRUE | TRUE | TRUE | Ask for confirmation for install | Show License
Yes | FALSE | FALSE | FALSE | Don't ask for confirmation | Show License
Yes | empty | FALSE | empty | Don't ask for confirmation | Show License
No | TRUE | TRUE | TRUE | Error. Can't generate nupkg | N/A
No | FALSE | FALSE | FALSE | Don't ask for confirmation | Don't Show license
No | empty | FALSE | empty | Don't ask for confirmation | Don't Show license

Also, Wee needto make sure that packages generated with this change can be installed in the following scenarios:

- Install in PM UI
- Install in PM Console
- Install in nuget.exe
- Install in dotnet.exe

As you can see, it is a lot of work and breaking changes. What we can do is to accept the new tests contained in this PR.

nkolev92 commented:
2 things that important from my perspective: 

* We should not introduce a binary breaking change
* We should not break the ability to round trip a nuspec.

@dominoFire has nicely summarized that in his table, specifically this PR currently changes Row 5.

If we can't promise the both my suggestion would be to not accept this PR and close the issue as won't fix. 

The first point is easier to avoid, so if we decide to take this PR as is we need a very good justification as to why we should break the second point.

jnm2 commented:
Thanks for these constraints. My next idea was to add a new bool property on `ManifestMetadata` such as `RequireLicenseAcceptanceSpecified` or `SuppressRequireLicenseAcceptance`. What are your thoughts on the viability of this approach?

jnm2 commented:
Since I believe the approach I mentioned could address both your points, I'll plan to implement it if I don't hear back from you soonish.

dominoFire commented:
> Since I believe the approach I mentioned could address both your points, I'll plan to implement it if I don't hear back from you soonish.

Apologies for the late response.

I think the approach is not viable. There's no direct access in `ManifestMetadata` to know if a field in the input nuspec is set. Take a look at `PackageBuilder`, that builds and saves the package to disk.

Please, consider all discussion feedback and constrains in this PR and in the issue.

If you are not making more progress on this PR, please close this PR and create and new one with the new approach.

jnm2 committed:
Stop generating <requireLicenseAcceptance> from csproj when unspecified

jnm2 committed:
Improve test perf by reusing the restored project

jnm2 commented:
@dominoFire No worries, I certainly know the feeling! 

I addressed outstanding comments, removed the implementation commit, and pushed a new one. This approach satisfies these two points:

> - We should not introduce a binary breaking change
> - We should not break the ability to round trip a nuspec.

There are a few other options if this one isn't palatable, but I'll wait to hear your thoughts.

jnm2 commented on src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs:
```diff
@@ -177,6 +177,12 @@ public bool RequireLicenseAcceptance
             set;
         }
 
+        public bool SuppressRequireLicenseAcceptance
```
Positive might be better, e.g. `EmitRequireLicenseAcceptance`

nkolev92 replied:
I think we should that too. 
The default is more natural there. 

cc @dominoFire do you agree?

jnm2 replied:
Let me know if you would like me to do this.

jnm2 replied:
@dominoFire I will go ahead with this unless you see a problem with it.

jnm2 replied:
This is done in the last commit.

dominoFire commented:
> @dominoFire No worries, I certainly know the feeling!
> 
> I addressed outstanding comments, removed the implementation commit, and pushed a new one. This approach satisfies these two points:
> 
> > * We should not introduce a binary breaking change
> > * We should not break the ability to round trip a nuspec.
> 
> There are a few other options if this one isn't palatable, but I'll wait to hear your thoughts.

@jnm2 Looks good. Make sure to add tests to all clients (nuget.exe, msbuild.exe, dotnet.exe --already present in this PR--)

jnm2 commented:
@dominoFire I wasn't able to find another place where an SDK csproj was being packed. Are you asking me to start a new MsbuildPackTaskTests next to MsbuildRestoreTaskTests? Same question for the nuget.exe client.

dominoFire commented:
> @dominoFire I wasn't able to find another place where an SDK csproj was being packed. Are you asking me to start a new MsbuildPackTaskTests next to MsbuildRestoreTaskTests? Same question for the nuget.exe client.

If you can create MsbuildPackTaskTests, that would be awesome. Otherwise, let me know if you need help with that. Below are the tests I think are required for this PR.

### Test files

- [x] dotnet.exe (tested in this PR): <br/> [test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs](https://github.com/nuget/nuget.client/blob/dev/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PackCommandTests.cs)
- [ ] MSBuild.exe (missing): <br/> test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildPackTaskTests.cs
- [ ] NuGet.exe (needs test): <br/> [test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs](https://github.com/nuget/nuget.client/blob/dev/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPackCommandTest.cs)
- NuGet in VS
  * ~PackageManager Console: No. You cannot pack with PMC~
  * ~PackageManager UI: No. You cannot pack with PM UI~
  * ~"Pack" context menu in .NET Core projects: No. That feature lives in another repo, which it seems to rely on MSBuild.~

jnm2 commented:
I'll get back into this tonight, hopefully.

dominoFire commented:
🔔

jnm2 commented:
Oops, didn't finish it. I'll try again today.

jnm2 committed:
Stop generating <requireLicenseAcceptance> from csproj with nuget.exe pack

jnm2 commented:
@dominoFire I wasn't sure what to make of this error from MSBuild in my MsbuildPackTaskTests attempt. Can you help?

```
Microsoft (R) Build Engine version 16.5.0-preview-19606-01+2bf8920d5 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
Build started 2020-06-17 19:41:54.
Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\solution.sln" on node 1 (Pack target(s)).
ValidateSolutionConfiguration:
  Building solution configuration "Debug|Any CPU".
Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\solution.sln" (1) is building "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\a\a.csproj" (2) on node 1 (Pack target(s)).
C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Microsoft.CSharp.CurrentVersion.targets(324,5): error MSB4019: The imported project "C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Roslyn\Microsoft.CSharp.Core.targets" was not found. Confirm that the expression in the Import declaration "C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Roslyn\Microsoft.CSharp.Core.targets" is correct, and that the file exists on disk. [C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\a\a.csproj]
Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\a\a.csproj" (Pack target(s)) -- FAILED.
Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\solution.sln" (Pack target(s)) -- FAILED.
```

dominoFire commented:
> @dominoFire I wasn't sure what to make of this error from MSBuild in my MsbuildTackTaskTests attempt. Can you help?
> 
> ```
> Microsoft (R) Build Engine version 16.5.0-preview-19606-01+2bf8920d5 for .NET Framework
> Copyright (C) Microsoft Corporation. All rights reserved.
> 
> Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
> Build started 2020-06-17 19:41:54.
> Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\solution.sln" on node 1 (Pack target(s)).
> ValidateSolutionConfiguration:
>   Building solution configuration "Debug|Any CPU".
> Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\solution.sln" (1) is building "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\a\a.csproj" (2) on node 1 (Pack target(s)).
> C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Microsoft.CSharp.CurrentVersion.targets(324,5): error MSB4019: The imported project "C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Roslyn\Microsoft.CSharp.Core.targets" was not found. Confirm that the expression in the Import declaration "C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Roslyn\Microsoft.CSharp.Core.targets" is correct, and that the file exists on disk. [C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\a\a.csproj]
> Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\a\a.csproj" (Pack target(s)) -- FAILED.
> Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\e4d4b670\50febc96\solution\solution.sln" (Pack target(s)) -- FAILED.
> ```

Apologies for the late response, @jnm2 

It seems to be a problem of a missing targets file that the test is not copying correctly. The file appears to be in the obj folder:

![image](https://user-images.githubusercontent.com/1192347/86964668-4c80ab00-c11b-11ea-92a2-39d05e51071c.png)

The possible workaround could be copying the file from a custom target defined in the Msbuild.Integration.Tests.csproj file, as shown below:

https://github.com/NuGet/NuGet.Client/blob/ae805c05455b2961154f1b84c30a881728bf0ae4/test/NuGet.Core.FuncTests/Msbuild.Integration.Test/Msbuild.Integration.Test.csproj#L36-L38

Please, make sure you’re working on top of the development branch to ease the final merging and to have the latest NuGet changes.

Here’s a script that could help you rebase your current branch with the changes on top of the dev branch. If you need help, let us know.

```powershell
# Stop the script if any error is encountered
$ErrorActionPreference = 'Stop'

# Stash current changes
git stash

# Get current branch
$currentBranch = git symbolic-ref HEAD --short

# Fetch dev branch
git fetch --all
git checkout dev
git pull

# Go back to the branch to rebase
git checkout $currentBranch

# Use this rebase if you don't expect conflicts, 
# If any, use Visual Studio Code to resolve conflicts
# See also https://code.visualstudio.com/docs/editor/versioncontrol#_merge-conflicts
git rebase dev --strategy=recursive --strategy-option=patience

# Use this rebase if you prefer changes from dev branch over your current branch.
# Use with caution as it can delete some of your changes
# Before using this command, make a backup branch if the rebasing fails
# git rebase dev --strategy=recursive --strategy-option=theirs

# Restore current changes
git stash pop
```

dominoFire commented:
Hi @jnm2

Hope all is well. 

Are you working on this PR? It's been a while since we hear from you. I hope you and your loved ones are doing well in these difficult times.

We would love to hear back from you. As stated in our [Contribution guide](https://github.com/NuGet/NuGet.Client/blob/dev/CONTRIBUTING.md#contributing-step-by-step), we will have to close this PR if there's no activity within 1 month.

Let us know if you need help.

jnm2 commented:
Thanks for letting me know about the time limit! I'll do your suggestions. It's nice to know that I'm welcome to rebase.

jnm2 commented:
@dominoFire I rebased onto the latest dev (using `git pull upstream dev -r`) and resolved conflicts. Then I fixed the bad Microsoft.CSharp.Core.targets destination like you pointed out. This got me past that hurdle. Next hurdle:

```
    Microsoft (R) Build Engine version 16.5.0-preview-19606-01+2bf8920d5 for .NET Framework
    Copyright (C) Microsoft Corporation. All rights reserved.
    
    Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
    Build started 2020-08-07 15:59:07.
    Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\bb380e09\6733bcf3\solution\solution.sln" on node 1 (Pack target(s)).
    ValidateSolutionConfiguration:
      Building solution configuration "Debug|Any CPU".
    Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\bb380e09\6733bcf3\solution\solution.sln" (1) is building "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\bb380e09\6733bcf3\solution\a\a.csproj" (2) on node 1 (Pack target(s)).
    C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\bb380e09\6733bcf3\solution\a\a.csproj : error MSB4057: The target "Pack" does not exist in the project.
    Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\bb380e09\6733bcf3\solution\a\a.csproj" (Pack target(s)) -- FAILED.
    Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\bb380e09\6733bcf3\solution\solution.sln" (Pack target(s)) -- FAILED.
    
    Build FAILED.
    
    "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\bb380e09\6733bcf3\solution\solution.sln" (Pack target) (1) ->
    "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\bb380e09\6733bcf3\solution\a\a.csproj" (Pack target) (2) ->
      C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\bb380e09\6733bcf3\solution\a\a.csproj : error MSB4057: The target "Pack" does not exist in the project.
    
        0 Warning(s)
        1 Error(s)
```

I assume more targets are missing since there is no file under `test\NuGet.Core.FuncTests\Msbuild.Integration.Test` containing `Name="Pack"`. How would I know where to find them?

I'm guessing I'll need to copy in `src\NuGet.Core\NuGet.Build.Tasks.Pack\NuGet.Build.Tasks.Pack.targets`, but how do I get it to be imported?

Looks like this is how it is imported for restore tests:
https://github.com/NuGet/NuGet.Client/blob/0b2568615d61e9f9393d597cb806f02af1b40a78/test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildIntegrationTestFixture.cs#L33-L39

However, the text `NuGetPackTargets` does not appear in any .targets file in the repo like it does for `NuGetRestoreTargets`, so there does not appear to be a path to mirror `/p:NuGetRestoreTargets`.

There's also ImportNuGetBuildTasksPackTargetsFromSdk in the code above, so it looks as though the Pack task was already intended to be usable with MsbuildINtegrationTestFixture.

jnm2 commented:
Ooh, just happened upon `/p:NuGetBuildTasksPackTargets` in a powershell script while scanning the repo for occurrences of `ImportNuGetBuildTasksPackTargetsFromSdk`

jnm2 commented:
Pushed a commit showing that this too appears to be a bit of a dead end. What am I missing?

nkolev92 commented on test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildPackTaskTests.cs:
```diff
@@ -0,0 +1,56 @@
+
+                var net461 = NuGetFramework.Parse("net461");
+
+                var projectA = new SimpleTestProjectContext(
```
The test project you are using is not SDK based.
Only SDK based projects import the pack SDK by default. Refer to https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package-msbuild. 

There is utility method in the simpletestprojectcontext CreateNETCoreWithSDK. 

Try using that instead

jnm2 commented:
@nkolev92 Thanks, updated. This is what happens now. Like always, not sure where to start to learn about this other than "find all files" attempts:

```
    Microsoft (R) Build Engine version 16.5.0-preview-19606-01+2bf8920d5 for .NET Framework
    Copyright (C) Microsoft Corporation. All rights reserved.
    
    Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
    Build started 2020-08-07 22:26:09.
    Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\solution.sln" on node 1 (Pack target(s)).
    ValidateSolutionConfiguration:
      Building solution configuration "Debug|Any CPU".
    Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\solution.sln" (1) is building "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj" (2) on node 1 (Pack target(s)).
    C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj(5,3): error MSB4247: Could not load SDK Resolver. A manifest file exists, but the path to the SDK Resolver DLL file could not be found. Manifest file path 'C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\SdkResolvers\Microsoft.Build.NuGetSdkResolver\Microsoft.Build.NuGetSdkResolver.xml'. SDK resolver path: C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Microsoft.Build.NuGetSdkResolver.dll
    Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj" (Pack target(s)) -- FAILED.
    Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\solution.sln" (Pack target(s)) -- FAILED.
    
    Build FAILED.
    
    "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\solution.sln" (Pack target) (1) ->
    "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj" (Pack target) (2) ->
      C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj(5,3): error MSB4247: Could not load SDK Resolver. A manifest file exists, but the path to the SDK Resolver DLL file could not be found. Manifest file path 'C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\SdkResolvers\Microsoft.Build.NuGetSdkResolver\Microsoft.Build.NuGetSdkResolver.xml'. SDK resolver path: C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Microsoft.Build.NuGetSdkResolver.dll
    
        0 Warning(s)
        1 Error(s)
```

dominoFire commented:
> @nkolev92 Thanks, updated. This is what happens now. Like always, not sure where to start to learn about this other than "find all files" attempts:
> 
> ```
>     Microsoft (R) Build Engine version 16.5.0-preview-19606-01+2bf8920d5 for .NET Framework
>     Copyright (C) Microsoft Corporation. All rights reserved.
>     
>     Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
>     Build started 2020-08-07 22:26:09.
>     Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\solution.sln" on node 1 (Pack target(s)).
>     ValidateSolutionConfiguration:
>       Building solution configuration "Debug|Any CPU".
>     Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\solution.sln" (1) is building "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj" (2) on node 1 (Pack target(s)).
>     C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj(5,3): error MSB4247: Could not load SDK Resolver. A manifest file exists, but the path to the SDK Resolver DLL file could not be found. Manifest file path 'C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\SdkResolvers\Microsoft.Build.NuGetSdkResolver\Microsoft.Build.NuGetSdkResolver.xml'. SDK resolver path: C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Microsoft.Build.NuGetSdkResolver.dll
>     Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj" (Pack target(s)) -- FAILED.
>     Done Building Project "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\solution.sln" (Pack target(s)) -- FAILED.
>     
>     Build FAILED.
>     
>     "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\solution.sln" (Pack target) (1) ->
>     "C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj" (Pack target) (2) ->
>       C:\Users\Joseph\Source\Repos\NuGet.Client\.test\work\0dcdb8aa\1a86c006\solution\a\a.csproj(5,3): error MSB4247: Could not load SDK Resolver. A manifest file exists, but the path to the SDK Resolver DLL file could not be found. Manifest file path 'C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\SdkResolvers\Microsoft.Build.NuGetSdkResolver\Microsoft.Build.NuGetSdkResolver.xml'. SDK resolver path: C:\Users\Joseph\Source\Repos\NuGet.Client\test\NuGet.Core.FuncTests\Msbuild.Integration.Test\bin\Debug\net472\win7-x64\Microsoft.Build.NuGetSdkResolver.dll
>     
>         0 Warning(s)
>         1 Error(s)
> ```

We need to update the MSBuild fixture to have NuGetSdkResolver (included in this codebase), the NuGet pack targets (included in this codebase) and the Microsoft.NET.Sdk to run properly.

I'm investigating now.

zivkan commented:
> We need to update the MSBuild fixture to have NuGetSdkResolver (included in this codebase), the NuGet pack targets (included in this codebase) and the Microsoft.NET.Sdk to run properly.

Our Dotnet.Integration.Tests fixture already does this.

My feedback is that it's more effort than it's probably worth for other test projects. If there's functionality that should be tested with NuGet.CommandLine.FuncTests or MSBuild.Integration.Tests, use non-SDK style projects. Limit tests using SDK style projects to Dotnet.Integration.Tests.

jnm2 commented:
> Limit tests using SDK style projects to Dotnet.Integration.Tests.

Dotnet.Integration.Test already done (https://github.com/NuGet/NuGet.Client/pull/3240/files#diff-0a8ea4346824f4b1f738f32446ce80ca), so shall I go ahead and force push to drop the [wip] commit?

dominoFire commented:
> > Limit tests using SDK style projects to Dotnet.Integration.Tests.
> 
> Dotnet.Integration.Test already done (https://github.com/NuGet/NuGet.Client/pull/3240/files#diff-0a8ea4346824f4b1f738f32446ce80ca), so shall I go ahead and force push to drop the [wip] commit?

Yeah, you can do that or you can skip the test and create a new issue for fixing it later. 

There are some failing tests:

### Test Case Title

PackCommand_PackLicense_SimpleExpression_StandardLicense
PackCommand_PackLicense_ComplexExpression_WithNonStandardLicense
PackCommand_PackLicense_LicenseUrlIsBeingDeprecated
PackCommand_PackLicense_BasicLicenseFileReadFileFromNupkg
PackCommand_PackLicense_PackBasicLicenseFile
PackCommand_PackLicense_LicenseInformationIsNotIncludedInTheSnupkg

jnm2 commented:
I force pushed to drop https://github.com/NuGet/NuGet.Client/commit/210a7b5282c9e67b50df70fb5540ab77e94bb79f and rebase onto dev.

@dominoFire The unrelated formatting commit you added (https://github.com/NuGet/NuGet.Client/commit/c10cbcabf37b285bb694fa468691191b6116dfb2) disappeared when I rebased onto dev because https://github.com/NuGet/NuGet.Client/pull/3633/files#diff-3b4f77571071babeebaa7e10b64c2526R797 made the same fix in the meantime.

Next I'll fix the tests.

dominoFire commented:
> I force pushed to drop [210a7b5](https://github.com/NuGet/NuGet.Client/commit/210a7b5282c9e67b50df70fb5540ab77e94bb79f) and rebase onto dev.
> 
> @dominoFire The unrelated formatting commit you added ([c10cbca](https://github.com/NuGet/NuGet.Client/commit/c10cbcabf37b285bb694fa468691191b6116dfb2)) disappeared when I rebased onto dev because https://github.com/NuGet/NuGet.Client/pull/3633/files#diff-3b4f77571071babeebaa7e10b64c2526R797 made the same fix in the meantime.
> 
> Next I'll fix the tests.

Good! We now have a dotnet tool to check source files formatting. 

We use the following command:

https://github.com/NuGet/NuGet.Client/blob/d1e99c970e52630a67e5b069dcdd537b027d48cf/build/vsts_build.yaml#L204

To apply formatter changes, run the command without `--check` flag. Just make sure to run the formatter before publishing changes.

jnm2 committed:
Switch to the positive

jnm2 commented:
I believe I added all the tests that ended up being asked for, and the entire test suite passes locally for me. (`git clean -xdf; .\configure.cmd; .\build.cmd`).

dominoFire commented on src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs:
```diff
@@ -177,6 +177,12 @@ public bool RequireLicenseAcceptance
             set;
         }
 
+        public bool EmitRequireLicenseAcceptance
```
Thanks for the update! Apologies for the late response.

We should add a condition that throws when a license is required but EmitRequireLicenseAcceptance is set to false. Otherwise. You can end up with packages requiring a license but no RequireLicenseAcceptance tag added.

See test below for PackageBuilder:

```csharp
[Theory]
           [InlineData(true, true, true)]
        // [InlineData(true, true, false)] // invalid: license is required
           [InlineData(true, false, false)]
        // [InlineData(true, false, true)] // invalid: license not required
        // [InlineData(false, true, true)] // invalid: license is required
           [InlineData(false, true, false)] // This case should throw
           [InlineData(false, false, false)]
        // [InlineData(false, false, true)] // invalid: license not required
        public void EmitLicenseAcceptance_Suceeds(bool emitRequireLicense, bool requireLicense, bool result)
        {
            var license = NuGetLicenseExpression.Parse("MIT");
            var list = new string[] { };
            var builder = new PackageBuilder()
            {
                Id = "test",
                Version = new NuGetVersion("0.0.1"),
                Authors = {
                    "TestAuthors"
                },
                Description = "Test package for EmitRequireLicenseAcceptance",
                EmitRequireLicenseAcceptance = emitRequireLicense,
                RequireLicenseAcceptance = requireLicense,
                LicenseMetadata = requireLicense ?
                    new LicenseMetadata(LicenseType.Expression, "MIT", license, list, LicenseMetadata.EmptyVersion):
                    null,
                DependencyGroups =
                {
                    new PackageDependencyGroup(
                        targetFramework: NuGetFramework.Parse("netstandard1.4"),
                        packages: new[]
                        {
                            new PackageDependency("another.dep", VersionRange.Parse("0.0.1"))
                        })
                }
            };

            using (var testDir = TestDirectory.Create())
            {
                var packagePath = Path.Combine(testDir, "test.0.0.1.nupkg");

                using (FileStream nupkgStream = File.Create(packagePath))
                {
                    builder.Save(nupkgStream);
                }

                var reader = new PackageArchiveReader(packagePath);

                Assert.Equal(result, reader.NuspecReader.GetRequireLicenseAcceptance());

                using (var nureader = new StreamReader(reader.GetNuspec()))
                {
                    var nuspecContent = nureader.ReadToEnd();

                    if (result)
                    {
                        Assert.Contains("<requireLicenseAcceptance>true</requireLicenseAcceptance>", nuspecContent);
                    }
                }
            }
        }
```

jnm2 replied:
@dominoFire Would this be something for PackageBuilderTest.cs or NuGetPackCommandTest.cs (where it looks like you copied from)?

dominoFire replied:
@jnm2 I think it is for PackageBuilderTests.cs, since it is testing PackageBuilder . Thanks for reaching out!

jnm2 replied:
Added.

dominoFire reviewed (commented):
Thanks for your contribution! I added some comments on PackageBuilder

dominoFire commented:
@jnm2 I dismissed previous staled review. Please, see comments in PackageBuilder

dominoFire commented:
@jnm2 It has been a while. Hope everything is fine. Are you planning on working on this PR ? We're very close to merge. If we don't hear back from you, the PR will be closed in 1 month.

jnm2 commented:
Sorry, I'm just very busy. I'm looking at it now.

jnm2 committed:
Validate that EmitRequireLicenseAcceptance is true when RequireLicenseAcceptance is true

jnm2 commented:
@dominoFire I fixed merge conflicts and added the test you asked for. Let me know if there's more I should do. Additionally, I'm happy to do any rebasing you'd like.

dominoFire commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageBuilderTest.cs:
```diff
@@ -2857,6 +2859,67 @@ public void PackageBuilderPreserveFileLastWriteTime()
+        [InlineData(true, false)]
+        [InlineData(false, true)]
+        [InlineData(false, false)]
+        public void EmitRequireLicenseAcceptance(bool emitRequireLicense, bool requireLicense)
```
Although I suggested created this test with all cases in the same method, It would be better to separate succeeding and failing cases in separate methods.

See https://github.com/NuGet/NuGet.Client/pull/3716#pullrequestreview-525229595

jnm2 replied:
Done. I separated out a further category as well (ShouldNotEmitElement).

dominoFire reviewed (approved):
Thank you for your contribution, @jnm2 ! The PR looks good. It only needs a rebase and a comment

jnm2 commented:
Rebased to fix merge conflicts.

dominoFire reviewed (approved):
Signing-off latest commit changes.

donnie-msft reviewed (approved):
🥳 Thanks for the contribution, and I'm really happy to see the new tests!

I didn't notice a resolution to the questions in this comment: https://github.com/NuGet/NuGet.Client/pull/3240#issuecomment-587878680
Especially, do we have an issue to update the docs?

dominoFire commented:
> 🥳 Thanks for the contribution, and I'm really happy to see the new tests!
> 
> I didn't notice a resolution to the questions in this comment: [#3240 (comment)](https://github.com/NuGet/NuGet.Client/pull/3240#issuecomment-587878680)
> Especially, do we have an issue to update the docs?

Issue is created: https://github.com/NuGet/docs.microsoft.com-nuget/issues/2298

I tested with new packages generated with this change on nuget.org, NuGet PM UI for VS and NuGet Package Explore and none of them shows any major visible issues.

dominoFire commented:
@jnm2 Thank you so much for your contribution to NuGet! 📦
